doc/SONiC_VRRP_HLD.md

```diff
@@ -0,0 +1,582 @@
+__Figure 1: VRRP Architecture__
+
+### 4.2.1 CONFIG_DB changes
+
+**VRRP_TABLE**
+
+Producer:  config manager 
+
+Consumer: VrrpMgr
+
+Description: New table that stores VRRP configuration for per interface + VRID. 
+
+Schema:
+
+```
+;New table
+;holds the VRRP configuration per interface and VRID
+
+key = VRRP_TABLE:interface_name:vrid
+                          ; Interface name string like Vlan1 or PortChannel002 or Ethernet4
+                          ; vrid is an integer
+; field = value
+vrid     = 1*3DIGIT       ; VRRP Instance Identifier
+vip      = ip_address     ; Virtual IP address. This is a list of IP addresses
+priority = vrrp_priority  ; Priority of VRRP instance
+adv_interval = 1*3DIGITS  ; Advertisement interval for VRRP. Default = 1sec
+state    = vrrp_state     ; String denoting the state of VRRP instance
+version  = vrrp_version   ; VRRP version. Value will always be 2 for this release
+pre_empt = "true"/"false" ; VRRP pre-emption is enabled? Default is True
+track_interface = track_interface ; List of interfaces tracked by a VRRP instance
+  <Interface_Name>|weight|<weight>; This is repeated for the configured tracking interfaces 
+```
+
+Example:- 
+
+**admin@sonic:~$ redis-cli -n 4 keys VRRP\***
+ 1) "VRRP|Vlan8|8"
+ 2) "VRRP|Vlan1|1"
+ 3) "VRRP|Vlan10|10"
+ 4) "VRRP|Vlan2|2"
+ 5) "VRRP|Vlan4|4"
+ 6) "VRRP|Vlan5|5"
+ 7) "VRRP|Vlan3|3"
+ 8) "VRRP|Vlan7|7"
+ 9) "VRRP|Vlan6|6"
+10) "VRRP|Vlan9|9"
+
+**admin@sonic:~$ redis-cli -n 4 HGETALL "VRRP|Vlan1|1"**
+ 1) "vrid"
+ 2) "1"
+ 3) "vip"
+ 4) "4.1.1.100,"
+ 5) "priority"
+ 6) "80"
+ 7) "adv_interval"
+ 8) "1"
+ 9) "state"
+10) ""
+11) "version"
+12) "2"
+13) "pre_empt"
+14) "True"
+15) "track_interface"
+16) "Ethernet7|weight|10,PortChannel001|weight|10,"
+
+
+
+**Example:- Entery with multiple virtual IPs**
+
+admin@sonic:~$ redis-cli -n 4 HGETALL "VRRP|Vlan1|1"
+ 1) "vrid"
+ 2) "1"
+ 3) "vip"
+ 4) "4.1.1.100,4.1.1.200,4.1.1.50,4.1.1.150,"
+ 5) "priority"
+ 6) "80"
+ 7) "adv_interval"
+ 8) "1"
+ 9) "state"
+10) ""
+11) "version"
+12) "2"
+13) "pre_empt"
+14) "True"
+15) "track_interface"
+16) "Ethernet7|weight|10,PortChannel001|weight|10,"
+
+### 4.2.2 APP_DB Changes
+
+**VRRP_TABLE**
+
+Producer:  VrrpSyncd 
+
+Consumer: VrrpOrch
+
+Description: This is a new table that contains VRRP state information. This entry will be added to APP_DB for each VIP for every Master VRRP instance. 
+
+Schema:
+
+```
+; New table
+; holds the VRRP state and VMAC information
+
+key = VRRP_TABLE:interface_name:vip:type
+    interface_name        ; interface name as a string. Vlan, Ethernet or PortChannel
+    vip                   ; virtual IP address in a.b.c.d/32 format
+    type                  ; IP address type string. only IPv4 supported currently
+    
+; field = value
+vmac = virtual_mac_address ; Virtual MAC address associated with VRRP instance
+```
+
+Example:- 
+
+**Key**: VRRP_TABLE:Vlan1000:[40.10.8.101/32:ipv4]
```

Comment:
why not use interface table? why can we extend intforch? why do we need a separate vrrporch?

Reply:
Intforch was not used to separate the use case. In VRRP multiple my IPs and my MACs (virtual RIF) needs to be programmed and it could so happen that there could be multiple sources (different VRRP instances on the same interface), generating same my IP.

As intforch cannot handle it as is, creating vrrporch module to handle the functionality separately.